doc/opgprof.1.in
.TH OPGPROF 1 "@DATE@" "oprofile @VERSION@"
.UC 4
.SH NAME
opgprof \- produce gprof-format profile data
.SH SYNOPSIS
.br
.B opgprof
[
.I options
]
[event name]
.SH DESCRIPTION

.B opgprof
outputs gprof-format profile data for a given binary image,
from an OProfile session.

.SH OPTIONS
.TP
.BI "--help / -? / --usage"
Show help message.
.br
.TP
.BI "--version / -v"
Show version.

.SH ENVIRONMENT
No special environment variables are recognised by opgprof.

.SH FILES
.TP
.I /var/lib/oprofile/samples/
The location of the generated sample files.

.SH VERSION
.TP
This man page is current for @PACKAGE@-@VERSION@.

.SH SEE ALSO
.BR oprofile(1)


ChangeLog
2003-07-05  Philippe Elie  <phil.el@wanadoo.fr>

	* daemon/opd_image.c:
	* libdb/db_insert.c: 64 bits arch compile fix, thanks to Falk Hueffner
	 <falk@debian.org>

2003-07-02  Philippe Elie  <phil.el@wanadoo.fr>

	* pp/opreport.cpp: honor options::show_header

2003-06-30  Graydon Hoare  <graydon@redhat.com>

	* libpp/op_header.cpp: Handle printing multiple bitmask values.

2003-06-19  John Levon  <levon@movementarian.org>

	* doc/Makefile.am: fix make distcheck

2003-06-18  John Levon  <levon@movementarian.org>

	* pp/opannotate_options.c: hide the unhandled
	--base-dir option

2003-06-18  John Levon  <levon@movementarian.org>

	* pp/Makefile.am: don't build opdiff

2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: don't accept "-r" - it's too easy to typo

2003-06-17  Will Cohen  <wcohen@redhat.com>

	* daemon/opd_cookie.h: Add s390 syscall value.
	
2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: don't accept "-r" - it's too easy to typo

2003-06-17  John Levon  <levon@movementarian.org>

	* doc/srcdoc/Doxyfile: removed

	* doc/srcdoc/Doxyfile.in:
	* configure.in: generate doxygen version nr. automatically

2003-06-17  Philippe Elie  <phil.el@wanadoo.fr>

	* libdb/db_insert.cpp: missing initialization of error message on
	  error path

2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: accept some short forms for common
	operations

2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: accept --foo blah as well as
	--foo=blah

2003-06-16  John Levon  <levon@movementarian.org>

	* utils/opcontrol: "opcontrol" will show usage
	instead of doing nothing

2003-06-16  Will Cohen  <wcohen@nc.rr.com>

	* libabi/abi.cpp:
	* libabi/op_import.cpp: Add required include for cassert.

2003-06-16  John Levon  <levon@movementarian.org>

	* events/x86-64.hammer.unit_masks: remove some
	bogus FIXMEs

2003-06-15  John Levon  <levon@movementarian.org>

	* README: fix autogen.sh line

2003-06-15  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: 64 bit fixes

2003-06-15  John Levon  <levon@movementarian.org>

	* utils/op_help.c: c89 fix

2003-06-15  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: c89 fix

2003-06-15  John Levon  <levon@movementarian.org>

	* doc/oprofile.xml: document --event

	* utils/opcontrol:
	* utils/op_help.c: allow unit mask, kernel, user to be
	optional. Verify not too many events are passed.

2003-06-15  John Levon  <levon@movementarian.org>

	* libop/op_cpu_type.c: fix nr_counters cpu type check

	* utils/op_help: implement --check_events, event validation
	and counter allocation

	* utils/opcontrol: use the above, implement --event=default

2003-06-14  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: fix a stupid error in my previous TGID patch

2003-06-12  Graydon Hoare  <graydon@redhat.com>

	* libop/op_events.c (op_check_events): Check bitmasks as well as 
	exclusive values, for unit masks.

2003-06-12  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: we were not handling a truncated
	TGID entry properly

2003-06-12  John Levon  <levon@movementarian.org>

	* libutil++/unique_storage.h: missing typename

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* pp/opannotate_options.cpp: typo in option name

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libutil++/op_bfd.cpp: boring_symbol() new to get a better choice
	when eliminatiing symbol at identical vma

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* utils/opcontrol: fix rtc option checking, bug added after 0.5.3,
	no big deal

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libregex/op_regex.cpp:
	* libpp/profile.cpp: compile fix, tree was broken by my previous patch

2003-06-11  Jason Yeh  <jason.yeh@amd.com>

	* events/i386.athlon.events: add CPU_CLK_UNHALTED

	* utils/opcontrol: use the above by default

	* events/x86-64.hammer.events:
	* events/x86-64.hammer.unit_masks: various fixes

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/op_header.cpp:
	* libpp/profile.cpp:
	* libpp/profile_spec.cpp:
	* libpp/symbol_sort.cpp:
	* libregex/op_regex.cpp:
	* libutil++/op_bfd.cpp: use exception rather exit(EXIT_FAILURE) in
	library code

2003-06-11  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: future-proof the code handlers

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* pp/opannotate.cpp: output_asm() avoid output when the set of selected
	symbols doesn't contain any samples

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libutil++/unique_storage.h: we don't need to check if key is present
	  before trying to insert it

2003-06-10  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/name_storage.h: minor tidy
	* libutil++/unique_storage.h: small efficiency improvement

2003-06-09  John Levon  <levon@movementarian.org>

	* pp/Makefile.am:
	* libutil/Makefile.am: fix make distcheck

2003-06-09  John Levon  <levon@movementarian.org>

	* libpp/opp_symbol.h: rename to ...

	* libpp/symbol.h: ... this

	* libpp/Makefile.am:
	* libpp/format_output.h:
	* libpp/format_output.cpp:
	* libpp/profile_container.h:
	* libpp/profile_container.cpp:
	* libpp/sample_container.h:
	* libpp/sample_container.cpp:
	* libpp/symbol_functors.h:
	* libpp/symbol_sort.cpp: fix up from above and some header
	cleanups

2003-06-09  John Levon  <levon@movementarian.org>

	* libpp/profile.h: std::pair not pair

2003-06-08  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/profile.cpp: don't underflow start offset when the vma range
	  is unknown
	* pp/opreport.cpp: use samples_range() not samples_range(0, ~0)

2003-06-08  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/profile.h:
	* libpp/profile.cpp: replace accumulate_samples() API by an iterator
	  interface
	* libpp/profile_container.h:
	* libpp/profilecontainer.cpp:
	* pp/opreport.cpp: use new API

2003-06-08  Philippe Elie  <phil.el@wanadoo.fr>

	* include/sstream: replace old version by the once shipped with

pp/opgprof.cpp
/**
 * @file opgprof.cpp
 * Implement opgprof utility
 *
 * @remark Copyright 2003 OProfile authors
 * @remark Read the file COPYING
 *
 * @author John Levon
 * @author Philippe Elie
 */

#include <iostream>
#include <cstdio>

#include "op_header.h"
#include "profile.h"
#include "op_libiberty.h"
#include "op_fileio.h"
#include "string_filter.h"
#include "file_manip.h"
#include "profile_container.h"
#include "partition_files.h"
#include "opgprof_options.h"
#include "cverb.h"
#include "ocg_hash.h"

using namespace std;

namespace {

#define GMON_VERSION 1
#define GMON_TAG_TIME_HIST 0
#define GMON_TAG_CG_ARC 1

struct gmon_hdr {
	char cookie[4];
	u32 version;
	u32 spare[3];
};


void op_write_vma(FILE * fp, op_bfd const & abfd, bfd_vma vma)
{
	// bfd vma write size is a per binary property not a bfd
	// configuration property
	switch (abfd.bfd_arch_bits_per_address()) {
		case 32:
			op_write_u32(fp, vma);
			break;
		case 64:
			op_write_u64(fp, vma);
			break;
		default:
			cerr << "oprofile: unknwon vma size for this binary\n";
			exit(EXIT_FAILURE);
	}
}


void get_vma_range(bfd_vma & min, bfd_vma & max,
                   profile_container const & samples)
{
	min = bfd_vma(-1);
	max = 0;

	sample_container::samples_iterator it  = samples.begin();
	sample_container::samples_iterator end = samples.end();
	for (; it != end ; ++it) {
		if (it->second.vma < min)
			min = it->second.vma;
		if (it->second.vma > max)
			max = it->second.vma;
	}

	if (min == bfd_vma(-1))
		min = 0;
	// we must return a range [min, max) not a range [min, max]
	if (max != 0)
		max += 1;
}


/**
 * @param abfd  bfd object
 * @param samples_files  profile container to act on
 * @param gap  a power of 2
 *
 * return true if all sample in samples_files are at least aligned on gap. This
 * function is used to get at runtime the right size of gprof bin size
 * reducing gmon.out on arch with fixed size instruction length
 *
 */
bool aligned_samples(profile_container const & samples, int gap)
{
	sample_container::samples_iterator it  = samples.begin();
	sample_container::samples_iterator end = samples.end();
	for (; it != end ; ++it) {
		if (it->second.vma % gap)
			return false;
	}

	return true;
}


void output_cg(FILE * fp, op_bfd const & abfd, samples_ocg_t const & cg_db)
{
	ocg_node_nr_t node_nr, pos;
	ocg_node_t * node = ocg_get_iterator(&cg_db, &node_nr);

	opd_header const & head = *static_cast<opd_header *>(cg_db.base_memory);
	bfd_vma offset = abfd.get_start_offset();
	if (!head.is_kernel)
		offset = 0;

	for (pos = 0; pos < node_nr; ++pos) {
		if (!node[pos].key)
			continue;

		bfd_vma from = node[pos].key >> 32;
		bfd_vma to = node[pos].key & 0xffffffff;

		from += offset;
		to += offset;
		op_write_u8(fp, GMON_TAG_CG_ARC);
		op_write_vma(fp, abfd, abfd.offset_to_pc(from));
		op_write_vma(fp, abfd, abfd.offset_to_pc(to));
		op_write_u32(fp, node[pos].value);
	}
}


void output_gprof(op_bfd const & abfd, profile_container const & samples,
                  samples_ocg_t const & cg_db, bool is_cg,
                  string const & gmon_filename)
{
	static gmon_hdr hdr = { { 'g', 'm', 'o', 'n' }, GMON_VERSION, {0,0,0,},};

	bfd_vma low_pc;
	bfd_vma high_pc;

	/* FIXME worth to try more multiplier ? is ia64 with its chunk of
	 * instructions can get sample inside a chunck or always at chunk
	 * boundary ? */
	int multiplier = 2;
	if (aligned_samples(samples, 4))
		multiplier = 8;

	cverb << "opgrof multiplier: " << multiplier << endl;

	get_vma_range(low_pc, high_pc, samples);

	cverb << "low_pc: " << hex << low_pc << " " << "high_pc: "
	      << high_pc << dec << endl;

	// round-down low_pc to ensure bin number is correct in the inner loop
	low_pc = (low_pc / multiplier) * multiplier;
	// round-up high_pc to ensure a correct histsize calculus
	high_pc = ((high_pc + multiplier - 1) / multiplier) * multiplier;

	cverb << "low_pc: " << hex << low_pc << " " << "high_pc: "
	      << high_pc << dec << endl;

	size_t histsize = (high_pc - low_pc) / multiplier;

	FILE * fp = op_open_file(gmon_filename.c_str(), "w");

	op_write_file(fp,&hdr, sizeof(gmon_hdr));
	op_write_u8(fp, GMON_TAG_TIME_HIST);

	op_write_vma(fp, abfd, low_pc);
	op_write_vma(fp, abfd, high_pc);
	/* size of histogram */
	op_write_u32(fp, histsize);
	/* profiling rate */
	op_write_u32(fp, 1);
	op_write_file(fp, "samples\0\0\0\0\0\0\0\0", 15);
	/* abbreviation */
	op_write_u8(fp, '1');

	u16 * hist = (u16*)xcalloc(histsize, sizeof(u16));

	profile_container::symbol_choice choice;
	choice.threshold = options::threshold;
	symbol_collection symbols = samples.select_symbols(choice);

	symbol_collection::const_iterator sit = symbols.begin();
	symbol_collection::const_iterator send = symbols.end();

	for (; sit != send; ++sit) {
		sample_container::samples_iterator it  = samples.begin(*sit);
		sample_container::samples_iterator end = samples.end(*sit);
		for (; it != end ; ++it) {
			u32 pos = (it->second.vma - low_pc) / multiplier;
			u32 count = it->second.count;

			if (pos >= histsize) {
				cerr << "Bogus histogram bin " << pos
			     << ", larger than " << pos << " !\n";
				continue;
			}
	
			if (hist[pos] + count > (u16)-1) {
				hist[pos] = (u16)-1;
				cerr <<	"Warning: capping sample count by "
				     << hist[pos] + count - ((u16)-1) << endl;
			} else {
				hist[pos] += (u16)count;
			}
		}
	}

	op_write_file(fp, hist, histsize * sizeof(u16));

	if (is_cg)
		output_cg(fp, abfd, cg_db);

	op_close_file(fp);

	free(hist);
}


string load_samples(op_bfd const & abfd, image_set const & images,
                    profile_container & samples)
{
	image_set::const_iterator it;
	for (it = images.begin(); it != images.end(); ) {
		pair<image_set::const_iterator, image_set::const_iterator>
			p_it = images.equal_range(it->first);

		string app_name = p_it.first->first;

		profile_t profile;

		for (;  it != p_it.second; ++it) {
			profile.add_sample_file(it->second.sample_filename,
						abfd.get_start_offset());
		}

		check_mtime(abfd.get_filename(), profile.get_header());

		samples.add(profile, abfd, app_name);
	}

	return images.begin()->first;
}


bool load_cg(samples_ocg_t & cg_db, image_set const & images)
{
	string const filename = images.begin()->second.sample_filename;

	string::size_type prefixend = filename.find_last_of("/");
	string const base = filename.substr(0, prefixend + 1);
	string const end = filename.substr(prefixend + 1, string::npos);

	string const cg_file = base + "{cg}/" + end;

	int rc = ocg_open(&cg_db, cg_file.c_str(), OCG_RDONLY,
		sizeof(struct opd_header));

	return rc == EXIT_SUCCESS;
}


int opgprof(vector<string> const & non_options)
{
	handle_options(non_options);

	profile_container samples(false, false, true);

	image_set images = sort_by_image(*sample_file_partition,
					 options::extra_found_images);

	// FIXME: symbol_filter would be allowed through option
	op_bfd abfd(images.begin()->first, string_filter());

	load_samples(abfd, images, samples);

	samples_ocg_t cg_db;

	bool const is_cg = load_cg(cg_db, images);

	output_gprof(abfd, samples, cg_db, is_cg, options::gmon_filename);

	if (is_cg)
		ocg_close(&cg_db);

	return 0;
}


} // anonymous namespace


int main(int argc, char const * argv[])
{
	return run_pp_tool(argc, argv, opgprof);
}
